File: scripts/qualify-split.mjs
// Gate for the near/far split: a giant with two authored moons inside a
// 512-body low-mass disk, integrated for 600 years by the production scheme
// (mutual tree far field at four coarse substeps, direct near field at eight
// fine steps each). Two comparisons: the same split with direct far forces
// isolates the tree's approximation, and a uniform sixteen-substep integration
// of the same tree forces isolates the integrator. The moons' elements are
// measured relative to their host with the same budgets as the small moon
// family; balances use the tree budgets.
import { readFile, writeFile } from 'node:fs/promises';
const raw = JSON.parse(await readFile(process.argv[2], 'utf8')),
  budgets = {
    axis_relative: 2e-3,
    eccentricity: 2e-3,
    phase_radians: 0.2,
    apsis_radians: 0.1,
    relative_energy: 2e-5,
    angular_residual: 1e-9,
    momentum_residual: 1e-9,
  },
  G = 39.47841760435743,
  failures = [];
function bodies(state) {
  const out = [];
  for (let i = 0; i < state.length; i += 5) out.push(state.slice(i, i + 5));
  return out;
}
function elements(body, host) {
  const rx = body[0] - host[0],
    ry = body[1] - host[1],
    vx = body[2] - host[2],
    vy = body[3] - host[3],
    d = Math.hypot(rx, ry),
    v2 = vx * vx + vy * vy,
    mu = G * (body[4] + host[4]),
    dot = rx * vx + ry * vy,
    ex = ((v2 - mu / d) * rx - dot * vx) / mu,
    ey = ((v2 - mu / d) * ry - dot * vy) / mu;
  return {
    axis: 1 / (2 / d - v2 / mu),
    eccentricity: Math.hypot(ex, ey),
    phase: Math.atan2(ry, rx),
    apsis: Math.atan2(ey, ex),
  };
}
const wrap = (a) => Math.abs(Math.atan2(Math.sin(a), Math.cos(a)));
function moonErrors(candidate, reference) {
  const c = bodies(candidate),
    r = bodies(reference),
    worst = { axis_relative: 0, eccentricity: 0, phase_radians: 0, apsis_radians: 0 };
  for (const index of raw.moons) {
    const a = elements(c[index], c[raw.host]),
      b = elements(r[index], r[raw.host]);
    worst.axis_relative = Math.max(
      worst.axis_relative,
      Math.abs(a.axis - b.axis) / Math.abs(b.axis),
    );
    worst.eccentricity = Math.max(worst.eccentricity, Math.abs(a.eccentricity - b.eccentricity));
    worst.phase_radians = Math.max(worst.phase_radians, wrap(a.phase - b.phase));
    if (a.eccentricity >= 1e-4 && b.eccentricity >= 1e-4)
      worst.apsis_radians = Math.max(worst.apsis_radians, wrap(a.apsis - b.apsis));
  }
  return worst;
}
const initial = raw.initial_balances,
  reference = raw.runs.find((run) => run.name === 'uniform-tree'),
  production = raw.runs.find((run) => run.name === 'split-tree');
const rows = raw.runs.map((run) => {
  const row = {
    name: run.name,
    years: run.samples.at(-1).year,
    relative_energy: Math.max(
      ...run.samples.map((s) => Math.abs((s.balances[0] - initial[0]) / initial[0])),
    ),
    angular_residual: Math.max(...run.samples.map((s) => Math.abs(s.balances[1] - initial[1]))),
    momentum_residual: Math.max(
      ...run.samples.map((s) => Math.hypot(s.balances[2] - initial[2], s.balances[3] - initial[3])),
    ),
    moonsBound: run.samples.every((s) => {
      const b = bodies(s.state);
      return raw.moons.every((index) => elements(b[index], b[raw.host]).axis > 0);
    }),
  };
  const against = run === production ? reference : production;
  const errors = run.samples.map((s, k) => moonErrors(s.state, against.samples[k].state));
  row.compared_with = against.name;
  row.moon_errors = Object.fromEntries(
    Object.keys(errors[0]).map((key) => [key, Math.max(...errors.map((e) => e[key]))]),
  );
  return row;
});
for (const row of rows) {
  if (row.years !== 600) failures.push(`${row.name}: incomplete horizon`);
  for (const key of ['relative_energy', 'angular_residual', 'momentum_residual'])
    if (row[key] > budgets[key]) failures.push(`${row.name}: ${key} ${row[key]} > ${budgets[key]}`);
  if (!row.moonsBound) failures.push(`${row.name}: a moon left its host`);
  for (const [key, value] of Object.entries(row.moon_errors))
    if (value > budgets[key])
      failures.push(`${row.name} vs ${row.compared_with}: moon ${key} ${value} > ${budgets[key]}`);
}
const report = {
  schema: 1,
  physics: raw.physics,
  bodies: raw.bodies,
  scope:
    'Giant with two authored moons inside a 512-body low-mass disk for 600 years: the production near/far split with the mutual tree, compared with the same split on direct far forces (tree approximation) and with a uniform sixteen-substep integration of the same tree forces (integrator). Moon elements are relative to the host.',
  budgets,
  rows,
  passed: !failures.length,
  failures,
};
await writeFile('split-qualification.json', JSON.stringify(report, null, 2) + '\n');
console.log(JSON.stringify(report, null, 2));
if (failures.length) process.exitCode = 1;

File: crates/sim/examples/qualify-split.rs
//! Near/far split qualification: a giant with two authored moons inside a
//! 512-body low-mass disk, 600 years. The production Wisdom–Holman split
//! (tree far field) is compared with the same split on direct far forces,
//! isolating the tree's approximation, and with an independent integrator: a
//! uniform eight-substep kick-drift-kick of the same tree forces. The disk leaves a gap around the
//! giant's orbit: the probe is collisionless, and a grain diving inside what
//! would be the giant's contact radius is an unresolved encounter in any
//! scheme, not a property of the split. Writes JSON to stdout, progress to
//! stderr. Arguments: an optional horizon in years (shorter for quick checks),
//! then optional plan names to run only those, so CI can run each plan on
//! its own machine; `scripts/qualify-split.mjs` merges the outputs.
use celestial_sim::{benchmark::OrbitProbe, *};
use serde_json::json;
fn balances(s: &[f64]) -> [f64; 4] {
    let mut out = [0.; 4];
    for (i, a) in s.chunks_exact(5).enumerate() {
        out[0] += 0.5 * a[4] * (a[2] * a[2] + a[3] * a[3]);
        out[1] += a[4] * (a[0] * a[3] - a[1] * a[2]);
        out[2] += a[4] * a[2];
        out[3] += a[4] * a[3];
        for b in s.chunks_exact(5).skip(i + 1) {
            out[0] -=
                G * a[4] * b[4] / ((a[0] - b[0]).powi(2) + (a[1] - b[1]).powi(2) + 1e-8).sqrt();
        }
    }
    out
}
fn main() {
    let mut w = World::new(Config {
        mission: None,
        ..Default::default()
    })
    .unwrap();
    let years: u32 = std::env::args()
        .nth(1)
        .and_then(|a| a.parse().ok())
        .unwrap_or(600);
    w.apply(Command::SeedSwarm {
        count: 800,
        disorder: 0.,
    })
    .unwrap();
    w.apply(Command::LaunchMass {
        kind: Kind::Giant,
        mass: 318.,
        radius: 3.,
        angle: 0.,
        speed: 1.,
    })
    .unwrap();
    let giant = w.bodies.last().unwrap().id;
    for (distance, speed) in [(0.05, 1.), (0.075, -1.)] {
        w.apply(Command::LaunchMoon {
            parent: giant,
            kind: Kind::Rocky,
            mass: 0.003,
            distance,
            angle: 1.,
            speed,
        })
        .unwrap();
    }
    // Keep the star, the giant, its moons and the first 508 grains outside a
    // gap of four Hill radii around the giant's orbit: 512 bodies, so the tree
    // and the split are selected. Low debris mass (16 Earth masses scaled by
    // a thousandth) isolates the split's own error, as in the tree lifetime gate.
    let star = w.bodies[0].pos;
    let mut kept = vec![];
    let mut grains = 0;
    for b in &w.bodies {
        if b.kind == Kind::Dust {
            let r = b.pos.minus(star).norm();
            if (2.2..=3.8).contains(&r) || grains >= 508 {
                continue;
            }
            grains += 1;
        }
        kept.push(b);
    }
    assert_eq!(kept.len(), 512, "the tree and the split must be selected");
    let initial: Vec<f64> = kept
        .iter()
        .flat_map(|b| {
            [
                b.pos.x,
                b.pos.y,
                b.vel.x,
                b.vel.y,
                if b.kind == Kind::Dust {
                    b.mass * 0.001
                } else {
                    b.mass
                },
            ]
        })
        .collect();
    let host = kept.iter().position(|b| b.id == giant).unwrap();
    let moons: Vec<usize> = kept
        .iter()
        .enumerate()
        .filter(|(_, b)| b.parent == Some(giant))
        .map(|(i, _)| i)
        .collect();
    assert_eq!(moons.len(), 2);
    // (name, exact far forces, uniform reference, reference substeps)
    let selected: Vec<String> = std::env::args().skip(2).collect();
    let plans: Vec<_> = [
        ("split-tree", false, false, 0u32),
        ("split-direct", true, false, 0),
        ("uniform-tree", false, true, 8),
    ]
    .into_iter()
    .filter(|(name, ..)| selected.is_empty() || selected.iter().any(|s| s == name))
    .collect();
    assert!(!plans.is_empty(), "unknown plan: {selected:?}");
    // One plan per process in CI and in the release script, which run them
    // concurrently; WebAssembly under WASI has no threads.
    let runs: Vec<_> = plans
        .iter()
        .map(|&(name, exact, uniform, substeps)| {
            let mut p = OrbitProbe::new(&initial, exact).unwrap();
            p.set_uniform(uniform);
            let mut samples = vec![];
            for year in 1..=years {
                if uniform {
                    p.advance_refined(512, substeps);
                } else {
                    p.advance(512);
                }
                if year % 25 == 0 || year == years {
                    let s = p.state();
                    samples.push(json!({"year":year,"balances":balances(&s),"state":s}));
                    eprintln!("{name}: {year} years");
                }
            }
            json!({"name":name,"exact":exact,"uniform":uniform,"substeps":substeps,"samples":samples})
        })
        .collect();
    println!(
        "{}",
        json!({"schema":1,"physics":checkpoint::PHYSICS_ID,"bodies":512,"years":years,"host":host,"moons":moons,"debris_earth_masses":0.016,"initial_balances":balances(&initial),"runs":runs})
    );
}
